Added stateful deployment for main

fank opened the pull request:
This pull request introduces support for deploying n8n as a StatefulSet in Kubernetes, enabling better handling of persistent volumes for multi-replica setups. It includes updates to the Helm chart, configuration files, and documentation to guide users on when and how to use StatefulSets instead of Deployments.

### StatefulSet Support

* **Documentation Updates**:
  - Added a new section in `README.md` explaining the differences between StatefulSets and Deployments, their use cases, and configuration requirements. Includes a warning about incompatibility with `persistence.existingClaim` when using StatefulSets.
  - Added an example configuration file (`examples/values_stateful.yaml`) for deploying n8n as a StatefulSet with individual persistent volumes per replica.
  - Updated `examples/README.md` to reference the new example configuration.

* **Helm Chart Enhancements**:
  - Introduced a new `main.useStatefulSet` flag in `values.yaml` to toggle between StatefulSet and Deployment.
  - Added a validation check in `_helpers.tpl` to ensure `persistence.existingClaim` is not used with StatefulSets.
  - Updated `deployment.yaml` to skip Deployment creation when `main.useStatefulSet` is true. [[1]](diffhunk://#diff-7181d4c035d0ad8de7b68726348d1a19a49f10c8d9ea0feceef3a725b79ee8c8R1) [[2]](diffhunk://#diff-7181d4c035d0ad8de7b68726348d1a19a49f10c8d9ea0feceef3a725b79ee8c8R125)
  - Created a new `statefulset.yaml` template for StatefulSet deployments, including support for volumeClaimTemplates and other StatefulSet-specific configurations.

* **Persistence Configuration**:
  - Modified `_helpers.tpl` to ensure dynamic PVC provisioning is only used for Deployments when `main.useStatefulSet` is false.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Optional StatefulSet deployment providing per-replica persistent volumes, with corresponding headless + regular service behavior and an example values file for multi-replica setups.
* **Bug Fixes**
  * Pre-deployment validation to prevent incompatible persistence + StatefulSet configurations.
* **Documentation**
  * Expanded deployment guide explaining Deployment vs StatefulSet, recommended production settings, and configuration guidance for multi-replica deployments.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

joshuabaird commented:
Thanks so much, @Fank!

fank committed:
feat(helm): add StatefulSet deployment support for n8n

This PR introduces StatefulSet support for n8n deployments, enabling better handling
of persistent storage when running multiple replicas.

## Key Changes

- Add new StatefulSet template with volumeClaimTemplates for individual PVCs per replica
- Add headless Service for StatefulSet DNS management and pod identity
- Add useStatefulSet flag to toggle between Deployment and StatefulSet modes
- Update PVC template to skip creation when using StatefulSet (handled by volumeClaimTemplates)
- Add validation to prevent incompatible persistence configurations (existingClaim with StatefulSet)
- Include example configuration file for StatefulSet deployments
- Add comprehensive documentation explaining when and how to use StatefulSets

## Implementation Details

- StatefulSet creates individual PVCs for each replica, preventing data corruption from shared volumes
- Headless service provides stable network identity for StatefulSet pods
- Regular service (-svc suffix) provides load balancing for external access
- Validation helper ensures users don't misconfigure persistence settings
- All templates include namespace field for proper scoping

## Use Cases

StatefulSets are recommended when:
- Running multiple n8n replicas that need individual persistent storage
- Each instance needs to maintain its own workflow data
- Implementing high availability with proper data isolation

For proper multi-replica operation, users should:
- Enable Redis/Valkey for session state management
- Use external database instead of SQLite
- Configure shared encryption key across all replicas

Fixes #186

fank commented:
@RoseSecurity any change for you to review this?

ronballesteros commented:
Hello, are there plans to merge this soon?

RoseSecurity commented on charts/n8n/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
 name: n8n
-version: 1.0.14
+version: 1.0.15
```
I think this functionality fits more into a minor release than a patch.

fank committed:
Merge upstream/main into fk_statefulmain

Resolve conflict in Chart.yaml: bump version to 2.1.0 (minor release
for StatefulSet feature as per review feedback)

fank committed:
feat(statefulset): add terminationGracePeriodSeconds support

Sync with upstream deployment.yaml change to support configurable
terminationGracePeriodSeconds via main.terminationGracePeriodSeconds

fank commented:
Merged upstream/main and applied changes:

- Bumped version to **2.1.0** (minor release for StatefulSet feature, as suggested by @RoseSecurity)
- Synced `terminationGracePeriodSeconds` support to statefulset.yaml (added in upstream for deployment.yaml)

fank committed:
fix: address CodeRabbit review feedback

- Fix initContainers indentation in statefulset.yaml (nindent 8)
- Convert values_stateful.yaml line endings from CRLF to LF

fank committed:
fix: address additional CodeRabbit review feedback

- Fix podLabels whitespace handling (use {{- for trim)
- Add missing volumes section for extraVolumes support
